### ISSUE TITLE
Callback issue Android

### ISSUE THREAD

**Issue by domhadfield** [dooboolab/react-native-iap#325]
### Version of react-native-iap
`"react-native": "0.55.2"`
`"react-native-iap": "^2.3.19"`

### Platforms you faced the error (IOS or Android or both?)
Android

### Expected behavior
No crash

### Actual behavior
Fatal crash

Stacktrace:
```
java.lang.RuntimeException Illegal callback invocation from native module. This callback type only permits a single invocation from native code. 
    CallbackImpl.java:28 com.facebook.react.bridge.CallbackImpl.invoke
    PromiseImpl.java:30 com.facebook.react.bridge.PromiseImpl.resolve
    RNIapModule.java:155 com.dooboolab.RNIap.RNIapModule$4.run
    RNIapModule.java:124 com.dooboolab.RNIap.RNIapModule$3.onBillingSetupFinished
    BillingClientImpl.java:903 com.android.billingclient.api.BillingClientImpl$BillingServiceConnection.onServiceConnected
    LoadedApk.java:1835 android.app.LoadedApk$ServiceDispatcher.doConnected
    LoadedApk.java:1864 android.app.LoadedApk$ServiceDispatcher$RunConnection.run
    Handler.java:809 android.os.Handler.handleCallback
    Handler.java:102 android.os.Handler.dispatchMessage
    Looper.java:166 android.os.Looper.loop
    ActivityThread.java:7377 android.app.ActivityThread.main
    Method.java:-2 java.lang.reflect.Method.invoke
    RuntimeInit.java:469 com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run
    ZygoteInit.java:963 com.android.internal.os.ZygoteInit.main
```

### Tested environment (Emulator? Real Device?)
Live

### Steps to reproduce the behavior
Unable to reproduce, have many reports from live app before rollback on crashlytics and bugsnag. App is always in the background when crash occurs.

### Code
Initialised on app start:
```
function* loadAvailableSubscriptions() {
  try {
    const productIds = yield select(getProductIds);
    const subscriptions = yield call(RNIap.getSubscriptions, productIds);
    yield put(setAvailableSubscriptions(subscriptions));
  } catch (error) {
    console.error('Error: loadAvailableSubscriptions', error);
    reportError({ error });
  }
}

export default function* initializeSubscription(): Saga<*> {
  try {
    yield call(RNIap.initConnection);
    yield call(loadAvailableSubscriptions);
    yield put(subscriptionInitialized());
  } catch (error) {
    console.error('Error: initializeSubscription', error);
    reportError({ error });
  }
}
```

Calling every time app state changes:
```
export function* appStateChangeWorker(nextAppState: AppState): Saga<*> {
  if (nextAppState === 'active') {
    yield call(RNIap.initConnection);
  } else if (nextAppState === 'background' || nextAppState === 'inactive') {
    yield call(RNIap.endConnection);
  }
}
```

**Comment by Pingou** [comment]
Same problem.

**Comment by dooboolab** [comment]
I think to figure out your case, you need to put log before each `function call` and try to find out where it is invoking methods. There may be some situations in your code which are not thread safe.

**Comment by pyankoff** [comment]
Having the same problem with live version of the app :(
Trying to make sure I init and close connection just once by counting invocations. Anything else I can try for thread safety?

**Comment by pyankoff** [comment]
@domhadfield thanks for suggestion, will do that too as a temporary fix. I'll also try closing connection on switching to background.

I've also managed to crash the app in background by force closing Play Store app while the IAP connection is open. Here is the comment I've got this idea from: https://github.com/googlesamples/android-play-billing/issues/45#issuecomment-324466519

I'm not sure how many users might actually close Play Store, but maybe there are similar circumstances causing the crash? @dooboolab can this help somehow?
